unix_integration/src/idprovider/kanidm.rs

```diff
@@ -223,13 +223,18 @@ impl IdProvider for KanidmProvider {
                         Err(IdpError::BadRequest)
                     }
                 }
-            } // For future when we have different auth combos/types.
-              /*
-                _ => {
-                    error!("invalid authentication request state");
-                    Err(IdpError::BadRequest)
-                }
-              */
+            }
+            (
+                AuthCredHandler::DeviceAuthorizationGrant,
+                PamAuthRequest::DeviceAuthorizationGrant { .. },
+            ) => {
+                error!("DeviceAuthorizationGrant not implemented!");
```

Comment:
Shouldn't this be calling something like self.client.read().idm_account_device_auth_grant_verify()? Or are you going to add that to the interface in a subsequent PR?

Reply:
I did that for Himmelblau. I wasn't planning to implement this for Kanidm. So yes, this would be done in a subsequent PR, but not necessarily by me.

Reply:
I'm confused how you did this in himmelblau? The structure of this code currently has the resolver passing the requests back-forth between the pam module and the idp client. How did you get this to work in himmelblau without this being implemented to make a call into the Idp client?

Reply:
Why would Himmelblau use `KanidmClient`? The 'client' backend of Himmelblau is a `PublicClientApplication` from the Microsoft Authentication Library (msal). I'm in the process of switching this backend to instead use a `ConfidentialClientApplication` to support device enrollment.
Himmelblau doesn't implement all the features of `KanidmClient` because it doesn't need them. Take a look at the source in https://github.com/openSUSE/himmelblau/blob/main/src/common/src/idprovider/himmelblau.rs and see for yourself.

Reply:
We've got a meeting scheduled for this week. If you can make that, then perhaps we can hash this out in that meeting?

Reply:
It's not using the kani client - this is the resolver calling the *trait* for a client implementation. Kanidm happens to implement that trait, as would himmelblau. So you need to update https://github.com/kanidm/kanidm/blob/master/unix_integration/src/idprovider/interface.rs with the DAG interaction traits, then each implementation can choose to use or ignore it.

Reply:
You said I should implement something like `idm_account_device_auth_grant_verify`, which I assumed meant something similar to functions like `idm_account_unix_cred_verify` implemented in the **struct KanidmClient**. Your initial comment was not clear that you meant I should implement something in the **IdProvider trait**.
So do I understand now, that you mean we should have a function for handling DAG in the IdProvider trait?

Reply:
The way I've implemented this in Himmelblau is that `unix_user_online_auth_step` currently both begins the DAG and completes the DAG, depending on the conditions. In Himmelblau, we can auth via username and password, and then we get a error response if the user requires MFA. If `unix_user_online_auth_step` receives this error response (REQUIRES_MFA), it returns:
```rust
                    return Ok(AuthResult::Next(AuthRequest::DeviceAuthorizationGrant {
                        data: resp.into(),
                    }));
```
Where `data` is the response from an `initiate_device_flow` to the client.
This request is forwarded back to pam, which sends the requested prompt and then calls back to the daemon and eventually leads us back to `unix_user_online_auth_step` again. `unix_user_online_auth_step` matches against the received `(cred_handler, pam_next_req)` and checks for a `DeviceAuthorizationGrant`, at which point it begins polling for a response from the server.

Reply:
> So do I understand now, that you mean we should have a function for handling DAG in the IdProvider trait? 

You know, I think what happened is I had confused myself. I had *completely* misread that you implemented this section in the Kanidm.rs provider, rather than the trait. I think I need to re-read this whole PR to work out why I confused myself so badly 🤣

Reply:
Okay I see the mistake I have made - I mistakenly thought this was in the resolver code so I was commenting on a problem that didn't exist. I'm so sorry. Your PR as it currently is is all good.